### PR TITLE
fix transform-typescript logic to remove definite fields

### LINKED ISSUE #12146: Typescript: fields defined with definite assignment assertion are removed with TS compiler flag `useDefineForClassFields` set
## Bug Report

**Current behavior**
If I have a field defined using `!`, for example:
```typescript
class Person {
  job?: number;
  age!: number;

  constructor() {
    
  }
}
```

`babel` using `preset-typescript` will transform it to this (REPL [link](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=true&loose=false&code_lz=MYGwhgzhAEAKCmAnCB7AdtA3gKGtAVigEYD8AXNGgK4C2RSA3LtGAObwCEF1dj2zwdBAAuiKsGEpEACgCUWZnmYBfbMqA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript%2Cenv&prettier=false&targets=&version=7.11.6&externalPlugins=)):

```javascript
class Person {
  constructor() {}
}
```

I think is the behavior currently being implemented, given [this](https://github.com/babel/babel/tree/main/packages/babel-plugin-transform-typescript/test/fixtures/class/uninitialized-definite-with-declare-enabled) test case.

This is the same behavior I observe in Typescript version < 3.7, see this playground [link](https://www.typescriptlang.org/play?useDefineForClassFields=true&target=99&ts=3.6.3#code/MYGwhgzhAEAKCmAnCB7AdtA3gWAFDWjTAFt4B+ALmggBdEBLNAcwG48DgALekAE0XhoAhFTQBXYgCMkbfNDBN4IwhOmJZ7aMHS1EY4DRSIAFAEosmgL55LQA)

But Typescript 3.7 comes with the new compiler flag `useDefineForClassFields` ([doc](https://devblogs.microsoft.com/typescript/announcing-typescript-3-7/#the-usedefineforclassfields-flag-and-the-declare-property-modifier)). With this option enabled, see this playground [link](https://www.typescriptlang.org/play?useDefineForClassFields=true&target=99&ts=3.7.5#code/MYGwhgzhAEAKCmAnCB7AdtA3gWAFDWjTAFt4AuaCAF0QEs0BzAbjwICsUAjAfgrQFdinJC3zQwDeAEI+g4YlGtowdNUT9gVFIgAUASixKCSgL54TQA), for the same input above, what Typescript generates is:
```javascript
"use strict";
class Person {
    job;
    age;
    constructor() {
    }
}
```

**Expected behavior**
I think I get the decision behind stripping out fields defined using definite assignment assertion, so conceptually, they are no different than fields defined using `declare`. However, for my use case, I would really like `babel` to be in sycn with Typescript here for the case where `useDefineForClassFields` is set to `true`, that is I want to fields to be present and initialized to `undefined`. I think I have seen exception made in the past for not stripping away `!` fields like in #8238. I wonder if we can add a new setting to `preset-typescript` or make another exception.

**Environment**
I thought the above REPLs and playgrounds should suffice, but in case you need a minimal repo, I prepared this [repo](https://github.com/akphi/config-tester). For this the important bits in my configs are:
```js
// tsconfig.json
{
  ...
  "useDefineForClassFields": true,
  ...
}

// babel settings (in webpack)

presets: [
      '@babel/preset-env',
      '@babel/preset-react',
      './dev/test-preset',
      ['@babel/preset-typescript', {
        onlyRemoveTypeImports: true,
        allowDeclareFields: true
      }],
    ]

// where `dev/test-preset` looks like the following 

plugins: [
      ['@babel/plugin-proposal-decorators', { legacy: true }],
      ['@babel/plugin-proposal-class-properties', { loose: true }],
    ],
```

**Additional context**
If this is appropriate, I want to mention my use case. I have some classes where the constructor is not a good place to initialize some of the fields, but I want to keep defining them using `!` because they should not be nullable (If only JS/TS support overloading constructors, this would be a non-problem). I'm also using `mobx`, which creates an observable wrapper around fields so the recommended place to set this is in the constructor, and `mobx` cannot really wrap fields that do not exist because `babel` strips away the field defined with `!` during build time. `mobx` used to work fine in the past because they use the decorator form, i.e. `@observable age!: number` which at the time worked because of #8238

Also, another totally coincidental thing is I used to have `@babel/plugin-proposal-class-properties` running before `preset-typescript` so it emits `!` fields and things work just fine, but recently, because I use `declare` field like `declare name: string`, I have to make sure `preset-typescript` runs before `@babel/plugin-proposal-class-properties`. So this problem surfaces, then on that topic, there's #10311

Anyhow, this has been a long report, please let me know what you think or what I can (maybe) help with since I'm still pretty inexperience. Thanks a lot!

### PR DESCRIPTION
**Pull request by akphi**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12146
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    |
| Any Dependency Changes?  |
| License                  | MIT

I made sure to separate the handling for 2 cases: fields defined with the keyword `declare` and fields defined using the definite assignment operator.

For the latter, when the flag `allowDeclareFields` is true (can be treated equivalent to Typescript `useDefineForClassFields = true`) we will not remove the field.

Also, since I separate the handling, I adjust the error message thrown by each case slightly when the fields in these 2 cases are given values.

Also, I added a test for `allowDeclareFields = false` to ensure old behavior works fine

@nicolo-ribaudo What you said about babel's `allowDeclareFields` is similar to Typescript's `useDefineForClassFields` is interesting. Should we update the doc for `transform-typescript` here?

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12149"></a>

### CODE REVIEW

**Comment by akphi** [comment]
@nicolo-ribaudo please, could you take a look when you have some time?

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

**Comment by nicolo-ribaudo** [comment]
>  Should we update the doc for transform-typescript here?

Do you want to open a PR to the `babel/website` repo to add it to this list? https://babeljs.io/docs/en/babel-plugin-transform-typescript#typescript-compiler-options

**Comment by akphi** [comment]
@nicolo-ribaudo sure thing! About this change, when do you think we can release it?

**Comment by nicolo-ribaudo** [comment]
We have a 2-:heavy_check_mark: policy. After another approval, we can merge this PR and release it in the first next version (7.11.8 or 7.12.0).

File: packages/babel-plugin-transform-typescript/src/index.js [hunk 1]
```diff
@@ -69,17 +69,26 @@ export default declare(
               `@babel/plugin-transform-typescript or @babel/preset-typescript is enabled.`,
           );
         }
-        if (node.definite || node.declare) {
+        if (node.declare) {
           if (node.value) {
             throw path.buildCodeFrameError(
-              `Definitely assigned fields and fields with the 'declare' modifier cannot` +
-                ` be initialized here, but only in the constructor`,
+              `Fields with the 'declare' modifier cannot be initialized here, but only in the constructor`,
```
Comment (existentialism) [inline packages/babel-plugin-transform-typescript/src/index.js:75]: While we're here:

```suggestion
              `Fields with a 'declare' modifier can only be initialized within the constructor`,
```

**Review by existentialism — APPROVED** [review]

**Comment by existentialism** [comment]
@akphi thanks!

**Comment by akphi** [comment]
@nicolo-ribaudo @existentialism Thanks! I was about the make the changes you suggest 😀 Not too familiar with the release cycle of babel myself, when do you have an estimate for when will this happen?

**Comment by existentialism** [comment]
Soon :)

**Pull request merged**